Store: Stats: Referrer: Add ability to search and select

psealock opened the pull request:
Allow user to select a referrer from
1. A select box if there are greater than 5 referrers
![refs-long](https://user-images.githubusercontent.com/1922453/37885216-797c6470-3110-11e8-9f36-1ad7d771257b.gif)

2. Quick links if there are less than 5 referrers
![refs-short](https://user-images.githubusercontent.com/1922453/37885208-73e62938-3110-11e8-8157-4a228fedaed2.gif)

### Testing
Head to `/store/stats/referrers/day/my-cool-site` and navigate to a date with referrers and test out the selection process. Check periods with greater than 5 and less than 5 referrers.

### Feeback Requested
How is the experience? Anything not working or could be improved? Should we handle a long list of search results with some kind of pagination?

greenafrican commented:
Really like the direction this is headed in. Two things I noticed (the screencast below is in real-time):

![search referrers](https://user-images.githubusercontent.com/1160732/37900115-85438d42-30ed-11e8-9a8b-fcea3529be9b.gif)

1. The list was empty until I selected the search box
2. I typed in google and was not sure what was happening as the characters didn't appear and then suddenly appeared. Perhaps a big lag as the list is getting filtered but it was pretty unusable for me.

psealock commented:
>The list was empty until I selected the search box

That was by design. Should be open by default if no referrer is selected?

>I typed in google and was not sure what was happening as the characters didn't appear and then suddenly appeared. Perhaps a big lag as the list is getting filtered but it was pretty unusable for me.

This data has 900 - 1400 referrers for any given day, so the performance is an issue with this much data. I think the bottle neck is in rendering, but I intend to do some testing to find out.

psealock commented:
### Rendering performance lag
The lag experienced in filtering with large amounts of data is directly related to the time required to paint each `<HorizontalBar>`. Using `PureComponent` or `shouldComponentUpdate` doesn't really help in this scenario because the `extent` prop is new on each filter. This is the mechanism that gathers the range (0 → max) to appropriately scale each block and is a function of the entire data array.

We could render a less computationally intensive component. Or...

### Render a max of 10 (5?) results with pseudo-pagination
![screen shot 2018-03-27 at 2 22 55 pm](https://user-images.githubusercontent.com/1922453/37941386-73fb6950-31ca-11e8-94f4-6efdabd4c269.png)


Search results, filtered or not, can be capped at a maximum length and paginated. 

This lends itself well to the concern raised earlier:

>The list was empty until I selected the search box

With pagination, the results could always be present. If there are more than the maximum per page the pagination buttons appear and a filter box. Maybe the filter box could even be below the results.

cc @greenafrican

psealock commented:
Compact mode for pagination pull request here, https://github.com/Automattic/wp-calypso/pull/23664

greenafrican commented:
> The lag experienced in filtering with large amounts of data is directly related to the time required to paint each `<HorizontalBar>`

What do you think about adding a search button, which is clicked once you've finished typing? Or perhaps a delay and/or a short period of inactivity to signal the end of typing?

psealock commented:
@greenafrican Can you give this a run through?

### Paginated Search
There may be a few kinks and bugs, but this should be enough to get a feel for it.
* There is always an expanded selection of at most 5
* More than 5 is paginated
* A direct link to a selected referrer via url query parameter automatically goes to correct page
* The correct page is visible after a selection is made and the filter is cleared
* Pagination is updated on clear

![refs](https://user-images.githubusercontent.com/1922453/38006954-b3a15cf2-32a3-11e8-832c-3be8fd69aba9.gif)

greenafrican commented:
Noticed that the filtered term disappears after a user navigates to and from the current date (which has the 'no data' notice):

![filter](https://user-images.githubusercontent.com/1160732/38029173-72f5ba44-3295-11e8-9dd5-29c1be719cab.gif)

greenafrican commented:
@psealock 

A few other notes:

- should we perhaps use the orange used when selecting dates rather than the green?
- I played with the layout and quite like the search component above the list and a list of 10 items rather than 5 (a teeny bit slower)
- how about excluding all $0 items? seems like might help the filters etc.

![screen shot 2018-03-28 at 15 10 39](https://user-images.githubusercontent.com/1160732/38030979-6bf1e074-329a-11e8-9a7a-00522d55b2a5.png)

psealock commented:
>the filtered term disappears after a user navigates to and from the current date

Good find, will address

>use the orange used when selecting dates rather than the green

Agree

>the search component above the list and a list of 10 items rather than 5

I can play around with this... its an easy modification

>how about excluding all $0 items? seems like might help the filters etc.

In many stores this could exclude a significant portion of referrers. It seems important to keep them, I think

psealock commented:
### Update
* Changed selected color to orange
* Fix disappearing search term on period change
* Moved Search above results
* Showing 10 results instead of 5

![screen shot 2018-03-29 at 4 18 07 pm](https://user-images.githubusercontent.com/1922453/38068308-ff3df5c0-336c-11e8-9876-2bde0020fd5e.png)

psealock commented:
@greenafrican how do you feel about the width or amount of space this occupies?

justinshreve commented:
> What do you think about adding a search button, which is clicked once you've finished typing? Or waiting for the Enter/Return key? Or perhaps a delay and/or a short period of inactivity to signal the end of typing?

Maybe a `debounce` on the search would help here?

greenafrican commented:
@psealock 

> how do you feel about the width or amount of space this occupies?

If we keep the $0 referrers and use up the available width, perhaps we can report on the `views`, `carts, `purchases` too, which may further the understanding of a $0 sales figure.

![screen shot 2018-03-29 at 12 06 20](https://user-images.githubusercontent.com/1160732/38083447-e8bb7282-3349-11e8-8c52-7421410dc703.png)

![screen shot 2018-03-29 at 12 09 02](https://user-images.githubusercontent.com/1160732/38083496-09487edc-334a-11e8-866d-c25ae20291d5.png)

greenafrican commented on client/extensions/woocommerce/app/store-stats/store-stats-referrer-widget/index.js:
```diff
@@ -64,25 +74,68 @@ class StoreStatsReferrerWidget extends Component {
-		const sortedAndTrimmedData = sortBySales( selectedData.data, 5 );
-		const extent = [ 0, d3Max( sortedAndTrimmedData.map( d => d.sales ) ) ];
+		const paginatedData = this.paginate( data );
+		const extent = [ 0, d3Max( paginatedData.map( d => d.sales ) ) ];
```
I wonder if we shouldn't be using the extent as it relates to `data` rather than `paginatedData`? As you page through it may feel as though a referrer on a lower page is comparable to one on a higher page (visually).

greenafrican replied:
This would include persisting the extent of ALL the data when searching/filtering the data.

timmyc commented on client/extensions/woocommerce/app/store-stats/referrers/index.js:
```diff
@@ -21,61 +21,165 @@ import StoreStatsPeriodNav from 'woocommerce/app/store-stats/store-stats-period-
+	};
+
+	onSearch = str => {
+		this.setData( this.props, str );
```
I agree with @justinshreve's comment that perhaps we should debounce this `onSearch` to prevent setting state multiple times while the search input is keyed in by users.

psealock replied:
Debounce is useful when you'd like to delay a request when the user is typing. Here, we have the data already. If the purpose of the debounce is to hide the delays in rendering elements, we'd still have to have a spinner or something to let the user know something is happening while we hold off on the computation, filtering and rendering. 

This feels wrong to me because we'd be introducing a delay in order to mask a perfermance issue. I'd rather address the issues itself

timmyc commented on client/extensions/woocommerce/app/store-stats/referrers/index.js:
```diff
@@ -21,61 +21,165 @@ import StoreStatsPeriodNav from 'woocommerce/app/store-stats/store-stats-period-
-	unit: PropTypes.oneOf( [ 'day', 'week', 'month', 'year' ] ),
-	slug: PropTypes.string,
-};
+	componentWillReceiveProps( nextProps ) {
```
Random preference - but I like having lifecycle methods declared near the top of the component. Not a written standard but seems to be the common practice in calypso.

timmyc commented on client/extensions/woocommerce/app/store-stats/referrers/index.js:
```diff
@@ -21,61 +21,165 @@ import StoreStatsPeriodNav from 'woocommerce/app/store-stats/store-stats-period-
+		} = this.state;
+		const unitSelectedDate = getUnitPeriod( selectedDate, unit );
+		const showSearch = unfilteredDataLength > LIMIT;
+		const title = `${ translate( 'Store Referrers' ) }${
```
Do you think we should use a [placeholder](https://github.com/Automattic/i18n-calypso#mixing-strings-and-markup) here in the translate call for the `queryParams.referrer`?

psealock replied:
I'm not following exactly. The `queryParams.referrer` shouldn't get translated and there's no markup mixed with strings. Can you explain? Cheers

timmyc commented on client/extensions/woocommerce/app/store-stats/referrers/index.js:
```diff
@@ -21,61 +21,165 @@ import StoreStatsPeriodNav from 'woocommerce/app/store-stats/store-stats-period-
+						</tbody>
+					</table>
+				) }
+				<JetpackColophon />
```
🌟

psealock commented on client/extensions/woocommerce/app/store-stats/store-stats-referrer-widget/index.js:
```diff
@@ -64,65 +74,152 @@ class StoreStatsReferrerWidget extends Component {
 }
 
-export default connect( ( state, { siteId, statType, query } ) => {
+// @TODO: Do something better with this selector
```
Yes, it is... good point. If created using the `createSelector` function, there are cache benefits for quicker lookups

https://github.com/Automattic/wp-calypso/blob/1a8a31b9d9a024f8e0fdd3f098d5ec8a4a5945d6/client/lib/create-selector/index.js#L83

psealock replied:
@greenafrican @justinshreve `sortBySales` seems to be used in every use of the referrer data. How about we handle sorting by sales to the server side?

psealock commented on client/extensions/woocommerce/components/d3/horizontal-bar/index.js:
```diff
@@ -35,7 +35,9 @@ const HorizontalBar = ( { className, data, extent, currency, height, width } ) =
-		const isOffsetText = xPos - ( text.node().getBoundingClientRect().width + textMargin ) <= 0;
+		const isOffsetText =
+			xPos + ( text.node().getBoundingClientRect().width + textMargin * 2 ) <= scale( extent[ 1 ] );
+
```
Something is not quite right here.

![screen shot 2018-04-03 at 1 09 36 pm](https://user-images.githubusercontent.com/1922453/38223743-82a4da38-3740-11e8-82fe-91298b35acff.png)

The previous way, for reference

![screen shot 2018-04-03 at 1 08 46 pm](https://user-images.githubusercontent.com/1922453/38223752-916b7496-3740-11e8-85fe-a6ef2d7e5ed3.png)

What's the motivation for referencing the the right edge as opposed to the left edge?

greenafrican replied:
> Something is not quite right here.

Good catch. My edit doesn't check that the text can fit inside the bar if it needs to.

> What's the motivation for referencing the the right edge as opposed to the left edge?

I think it looks and reads better.

psealock commented on client/extensions/woocommerce/app/store-stats/store-stats-referrer-widget/index.js:
```diff
@@ -158,17 +162,32 @@ class StoreStatsReferrerWidget extends Component {
+								</TableItem>
+								<TableItem>
+									<HorizontalBar
+										extent={ extent_views }
```
I see that you are using the views extent so that the purchase events can be easily compared by size to the views block. But it begins to break down after a while. Over smaller data sets, this isn't as much of an issue, but i'm wondering about having the extent over the entire dataset as opposed to just the page in question. Does the ability to view from the perspective of the overall scale take too much away from having nice blocks?

![screen shot 2018-04-03 at 1 20 09 pm](https://user-images.githubusercontent.com/1922453/38224001-00cdcacc-3742-11e8-9d40-e90d0d31e861.png)

greenafrican replied:
The more I think about this the more I feel it was a mistake to include `views`, `carts` and `purchases` in this widget with all the confusing scales to interpret.

They're important stats but we can add another widget alongside or below the gross sales referrer widget and in another PR.

psealock commented:
Thanks @shaunandrews for the review!

>I think the bar graph is going to be problematic, especially on mobile.

Agree, we'll need to solve this problem

>I think for a lot of people, they'll end up with a list that looks something like this:

With more information associated with the `$0.00` values, it should look less sad

>you get some strange double arrows

I was wondering about this. I will pull this out to a separate issue because its a characteristic of the navigation for the entire of Store Stats

>you might notice that we already have a screen that lists referrer in Stats

We call it "Store Referrers" v. simply "Referrers". This can definitely be a source of confusion. The data is calculated separately and we need to do some thinking about how the two interact and differ.

psealock commented:
### Update
* Simplified the `filter` argument 1c491e4171ce1725e404a6eda75fbc4187374fe1
* Move lifecycle functions acd1711c89cffc5346853b0e9f256f0a3daa59fa
* Created a `getStoreReferrers` selector 4ee450d72925676b477dd0721ac7640fb714cec4

Thanks for the additions and reviews @greenafrican @timmyc @shaunandrews

psealock commented:
This one seems ready to go. Here is a list of unaddressed issues
* **Mobile view** - We've removed view, carts, and purchases, so there is more room there
* **Little data** - make it look less sad. Lets address this in a following PR
* **Double arrows** - Tracking here, https://github.com/Automattic/wp-calypso/issues/23839
* **Sort by sales server side** - lets keep this in mind and make an adjustment later

kellychoffman commented:
This doesn't feel good to go from a UX perspective as Shaun brought up here: https://github.com/Automattic/wp-calypso/pull/23579#issuecomment-377360668. I don't think we should ship this PR until these are addressed. 

I'll take a deeper look by the end of the day and respond in further detail.

kellychoffman commented:
> Test live: https://calypso.live/?branch=add/stats-referrer-search

@psealock: can you confirm this works to test out the branch? I'm not seeing any referral stats in /Store (but I do see referral stats in /Traffic)

kellychoffman commented:
My biggest concern is with the bar graph. It don't think its very useful and feels forced. Is there a reason to have them at all? Why not just list the numbers like we do in the other stats widgets below? (Popular Products, Top Categories, Coupons).
And @shaunandrews - your dot idea sounds interesting, but I can't quite picture it. Care to draw it up? 😀

Other concerns: 
- The double arrow/navigation issue mentioned here: https://github.com/Automattic/wp-calypso/pull/23579#issuecomment-377360668 (Agree its fine in another PR since its also happening in production) It should use the SectionNav to be consistent with other stats pages. 
- Placement on the page: Is this the most important stat on the page? It feels like it belong below the sparkline stats as its more of a detailed level stat.

psealock commented:
>My biggest concern is with the bar graph. It don't think its very useful and feels forced. Is there a reason to have them at all? Why not just list the numbers like we do in the other stats widgets below? (Popular Products, Top Categories, Coupons).

Are you speaking about the widget on the main store stats page? If so, keep in mind that the referrer widget will be smaller as the layout will change just before referrers goes live. The reason is because I can't easily change the layout and have our feature flag in place.

On the all referrers page `/store/stats/referrers/week/` the point is a mechanism for selecting a particular referrer to drill down into. In that respect, the bars aren't needed but I thought it was a good opportunity to visualize how one compares to others when selecting which one to take a closer look at.

>Placement on the page: Is this the most important stat on the page? It feels like it belong below the sparkline stats as its more of a detailed level stat.

Definitely. We are planning on removing two sparklines and having it occupy 50% (remaining 4 sparklines are the other 50%). Again, we're leaving this until later so the feature flag can easily be implemented.

psealock commented:
>Why not just list the numbers like we do in the other stats widgets below? 

So on the main stats page, it would look like this

![screen shot 2018-04-05 at 11 00 30 am](https://user-images.githubusercontent.com/1922453/38339120-c8314970-38c0-11e8-98d9-50a9800b76ff.png)

And on the All Referrers page, the search/selection would be this

![screen shot 2018-04-05 at 10 59 58 am](https://user-images.githubusercontent.com/1922453/38339083-886d50cc-38c0-11e8-980d-8ef09d2f9d3a.png)

@kellychoffman, Is this what you are suggesting?

psealock commented:
For reference, here is a (very) rough idea of what the page will look like once the main chart is added. Further widgets (such as conversions) will be added at the bottom. 

![screen shot 2018-04-05 at 2 56 31 pm](https://user-images.githubusercontent.com/1922453/38345185-c87b27ae-38e1-11e8-8662-ed25b33019f4.png)

kellychoffman commented:
Re:  https://github.com/Automattic/wp-calypso/pull/23579#issuecomment-378806660 

> For reference, here is a (very) rough idea of what the page will look like once the main chart is added. Further widgets (such as conversions) will be added at the bottom.

- The main chart is great. Let's def do that.
- The bar chart below seems unnecessary, it also hides the source which is problematic. Let's do the numbered chart you showed in https://github.com/Automattic/wp-calypso/pull/23579#issuecomment-378772386 which is clear, to the point, perfecto.

kellychoffman commented:
One question for clarification: Is "Filter referrers" only shown for sites with many referrers? It feels overboard when you have 10 or less.

psealock commented:
>One question for clarification: Is "Filter referrers" only shown for sites with many referrers? It feels overboard when you have 10 or less.

The search box is only shown if there are more than 10

>The bar chart below seems unnecessary, it also hides the source which is problematic. Let's replace that with the numbered chart you showed in #23579 (comment) which is clear, to the point, perfecto. (on the main stats page too)

I see the simplicity in it, but I'd love to hear another opinion before removing. @jameskoster @timmyc @greenafrican @justinshreve care to weigh in?

kellychoffman commented:
This is also inline with competitor's stats offerings. To name a couple:
- Squarespace has an overview graph + breakdown table (text only, no bar graph) https://support.squarespace.com/hc/en-us/articles/217551548-Traffic-Sources-Analytics
- Shopify does the same. Overview graph + breakdown list (text only, no bar graph)
https://pavetoolinnovators.myshopify.com/admin/reports/visits_over_time

timmyc commented:
This is an excellent topic, and one that I kind of feel has been explored before, but I'd have to do some serious slack/mgs/old claypso repo crawling to find it. 

The current standard for stats widgets has been to just display the raw data, and offer the csv export link for more in-depth excel gymnastics to take place offline. There are a few summary pages that do have the chart up top, which i think is a great pattern: viz up top, with raw data below.

I do echo Kelly's concern that the bars can be distracting to a point where the obscure some of the most critical data we are trying to portray - the referral source and the $ total for that source. So I'm leaning towards omitting the in-line bar graph that is shown in the tabular data myself.

One nice compromise here might be to introduce a "chart" action on the stat widget itself. So the default view would be the tabular data like we show in all widgets, then maybe in the header area of the widget, a chart icon is exposed, and when clicking on that, the data is portrayed in a viz - or in some other manner that conveys %'s. Just a random idea, but could be best of both worlds sort of thing.

Users are *very opinionated* on this matter. Some loathe visualizations, others love them. Some only want the data.... and I think we all fall into one of those two sides on this discussion too. People are just so opinionated with their data consumption 😆 

But my vote is for the raw data right now. I'm /cc'ing @LevinMedia for his thoughts too from a design perspective.

kellychoffman commented:
> One nice compromise here might be to introduce a "chart" action on the stat widget itself. So the default view would be the tabular data like we show in all widgets, then maybe in the header area of the widget, a chart icon is exposed, and when clicking on that, the data is portrayed in a viz - or in some other manner that conveys %'s. Just a random idea, but could be best of both worlds sort of thing.

I think this would be solved by the to portion of the screenshot here: https://github.com/Automattic/wp-calypso/pull/23579#issuecomment-378806660 

I worry that anything further would be overcomplicating this idea for a v1.

shaunandrews commented:
> I see the simplicity in it, but I'd love to hear another opinion before removing.

My $0.02 is that the bar chart is distracting and not any more helpful than the simple table. In fact, the bar graph is unhelpful in ways; It obscures the source URL (which, is like, the entire point here...) and it conflicts visually with the chart above it (vertical v. horizontal is weird, too).

psealock commented:
### Update
* Horizontal bars are removed in widget on main page and referrers page
![screen shot 2018-04-09 at 5 03 31 pm](https://user-images.githubusercontent.com/1922453/38480274-cd318592-3c18-11e8-8af5-8896ac52406a.png)
![screen shot 2018-04-09 at 5 00 42 pm](https://user-images.githubusercontent.com/1922453/38480275-cfb3bda8-3c18-11e8-9d6c-913f8e3b40f7.png)
* Selection indicated by `$blue-wordpress` background
* Rebased against master

kellychoffman commented:
Awesome. Just one question: I'm seeing a data show up below the card; is that intentional? 

<img width="890" alt="screen shot 2018-04-09 at 3 31 12 pm" src="https://user-images.githubusercontent.com/5835847/38524246-314efad6-3c0b-11e8-9ba9-20f1e9d4f1ee.png">

psealock commented:
Hi @kellychoffman, this is me just printing out the raw data of the selected referrer for dev purposes. We're still behind the feature flag and it gets replaced in the next PR by the chart. I'll try to remove that stuff to avoid confusion next time

kellychoffman commented:
Cool. Figured as such. Just didn't want to assume.
